Sources/DeepLearning/Layer.swift

```diff
@@ -268,3 +270,94 @@ public struct AvgPool2D<Scalar>: Layer
+public struct LayerNorm<Scalar>: Layer
+    where Scalar : BinaryFloatingPoint & Differentiable & TensorFlowScalar {
+    /// The offset value, also known as beta.
+    public var offset: Tensor<Scalar>
+
+    /// The scale value, also known as gamma.
+    public var scale: Tensor<Scalar>
+
+    @noDerivative public let axis: Int32
+
+    /// The variance epsilon value.
+    @noDerivative public let epsilon: Tensor<Scalar>
+
+    public init(featureCount: Int,
+                axis: Int,
+                epsilon: Tensor<Scalar> = Tensor(0.001)) {
+        self.scale = Tensor<Scalar>(ones: [Int32(featureCount)])
+        self.offset = Tensor<Scalar>(zeros: [Int32(featureCount)])
+        self.axis = Int32(axis)
+        self.epsilon = epsilon
+    }
+
+    @differentiable(wrt: (self, input))
+    public func applied(to input: Tensor<Scalar>) -> Tensor<Scalar> {
+        let mean = input.mean(alongAxes: axis)
+        let variance = input.variance(alongAxes: axis)
+        let inv = rsqrt(variance + epsilon) * scale
+        return (input - mean) * inv + offset
+    }
+}
+
+public extension Tensor where Scalar : BinaryFloatingPoint,
+                       Scalar.RawSignificand : FixedWidthInteger {
+    @differentiable(wrt: (self) where Scalar : Differentiable)
+    func droppedOut(withProbability dropProbability: Double) -> Tensor {
+        let noise = Tensor(randomUniform: shape)
+        let keepMask = noise .>= Scalar(dropProbability)
+        let keepProbability = Scalar(1.0 - dropProbability)
+        return self * Tensor(keepMask) / Tensor(keepProbability)
+    }
+}
+
+@_fixed_layout
+public struct Dropout<Scalar>: Layer
+    where Scalar: BinaryFloatingPoint & Differentiable & TensorFlowScalar,
+          Scalar.RawSignificand : FixedWidthInteger {
+    @noDerivative public let dropProb: Double
+    @noDerivative public let mode: ModeRef
+    var _unused: Tensor<Scalar>
+
+    public init(dropProbability: Double, modeRef: ModeRef) {
+        self.dropProb = dropProbability
+        self.mode = modeRef
+        self._unused = Tensor<Scalar>(Scalar(0))
+    }
+
+    @differentiable(wrt: (self, input))
+    public func applyTraining(to input: Tensor<Scalar>) -> Tensor<Scalar> {
+        return input.droppedOut(withProbability: dropProb)
+    }
+
+    @differentiable(wrt: (self, input))
+    public func applyInference(to input: Tensor<Scalar>) -> Tensor<Scalar> {
+        return input
+    }
+
+    @differentiable(wrt: (self, input), vjp: _vjpApplied(to:))
+    public func applied(to input: Tensor<Scalar>) -> Tensor<Scalar> {
+        if mode.training {
+            return applyTraining(to: input)
+        } else {
+            return applyInference(to: input)
+        }
+    }
+
+    public func _vjpApplied(to input: Tensor<Scalar>) ->
+        (Tensor<Scalar>, (Tensor<Scalar>) ->
+            (Dropout<Scalar>.CotangentVector, Tensor<Scalar>)) {
+        if mode.training {
+            return Swift.valueWithPullback(at: self, input) {
```

Comment:
Wonder if the [method-style differential operator](https://github.com/apple/swift/blob/a6a2790d64f8512e93e4ca75b0c1f2e0d70d48f3/stdlib/public/core/AutoDiff.swift#L215) works?
```swift
        if mode.training {
            return valueWithPullback(at: input) {
                $0.applyTraining(to: $1)
            }
        else {
            return valueWithPullback(at: input) {
                $0.applyInference(to: $1)
            }
        }
```

Reply:
It didn't when I first tried it, because of a mismatch between curried and noncurried signatures. Can try again though!

Reply:
Seems to work now. I was probably doing something wrong before.